Frameworks/Foundation/NSUndoManager.mm

```diff
@@ -16,16 +16,69 @@
 
 #import <StubReturn.h>
 #import <Foundation/NSUndoManager.h>
+#import <Foundation/NSMutableArray.h>
+#import "_NSUndoCall.h"
+#import "_NSUndoGroup.h"
 
-NSString* const NSUndoManagerGroupIsDiscardableKey = @"NSUndoManagerGroupIsDiscardableKey";
+@implementation NSUndoManager {
+    NSMutableArray<_NSUndoGroup*>* _undoStack;
+    NSMutableArray<_NSUndoGroup*>* _redoStack;
+}
 
-@implementation NSUndoManager
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
+*/
+- (id)init {
+    if (self = [super init]) {
+        _groupsByEvent = YES;
+        _NSUndoGroup* startingGroup = [[[_NSUndoGroup alloc] initWithOwner:self] autorelease];
+        _NSUndoGroup* startingRedoGroup = [[[_NSUndoGroup alloc] initWithOwner:self] autorelease];
+
+        _undoStack = [NSMutableArray array];
+        _redoStack = [NSMutableArray array];
+
+        [_undoStack addObject:startingGroup];
+        [_redoStack addObject:startingRedoGroup];
+
+        _undoRegistrationEnabled = YES;
+    }
+    return self;
+}
+
+/**
+ @Status Interoperable
 */
 - (void)registerUndoWithTarget:(id)target selector:(SEL)aSelector object:(id)anObject {
-    UNIMPLEMENTED();
+    if (!(self.isUndoRegistrationEnabled)) {
+        return;
+    }
+
+    _NSUndoCall* undoCall = [[[_NSUndoCall alloc] _initWithTarget:target selector:aSelector object:anObject] autorelease];
+    [self _registerUndoAction:undoCall];
+}
+
+// Internal function for proper registration of the deferred call.
+- (void)_registerUndoAction:(_NSUndoCall*)undoCall {
+    // If groupsByEvent is set then have a top level undo group to catch all undo calls.
+    if (_groupsByEvent) {
+        if (self.isUndoing) {
+            if ([_redoStack count] == 0) {
```

Comment:
It's a strange situation. As called out in the design doc, a top level group will be created if _groupsByEvent is true. It will also automatically close the top level group when an undo message is sent.

---
In reply to: [75724410](https://github.com/Microsoft/WinObjC/pull/900#discussion_r75724410) [](ancestors = 75724410)